File: src/Systems/Fortran/patterns/interpreter.f90
module interpreter_example
  implicit none
contains
  recursive integer function eval(kind, left, right) result(value)
    integer, intent(in) :: kind, left, right
    select case (kind)
    case (0); value = left
    case (1); value = left + right
    case default; value = left * right
    end select
  end function
  logical function run()
    run = eval(1, 7, eval(2, 3, 4)) == 19
  end function
end module

File: src/Systems/Swift/patterns/Interpreter.swift
enum InterpreterExample {
    indirect enum Expr {
        case lit(Int)
        case add(Expr, Expr)
        case mul(Expr, Expr)
    }

    static func evaluate(_ expression: Expr) -> Int {
        switch expression {
        case .lit(let value): return value
        case .add(let left, let right): return evaluate(left) + evaluate(right)
        case .mul(let left, let right): return evaluate(left) * evaluate(right)
        }
    }

    static func run() -> Bool {
        evaluate(.add(.lit(7), .mul(.lit(3), .lit(4)))) == 19
    }
}
